### DISCUSSION TITLE
How use request.sid

### DISCUSSION THREAD

**Discussion by t-h2o** [miguelgrinberg/Flask-SocketIO#2115] [Q&A]
# Hi! does `request.sid` work?

> My entire project and the version I had the problem is available [here](https://github.com/t-h2o/matcha/blob/discussions-flask-socketio-request-sid/backend/matcha/__init__.py)

## example use `request.sid`

from: https://github.com/miguelgrinberg/Flask-SocketIO/blob/e72896597503d6e9bc4c213f123bee8545dab614/example/app.py#L122

### work

The client receive a SID :)

```json
{"sid":"fNHMvLB-E6xLWKWdAAAB"}
```

### patch applied

```patch
-from flask import Flask
+from flask import Flask, request
 from flask_socketio import SocketIO, disconnect
 from flask_cors import CORS
 from os import environ
@@ -75,7 +75,7 @@ def create_app():

     @socketio.on("connect")
     def handle_connect():
-        sid = socketio.server.eio.sid
+        sid = request.sid
         user_id = verify_token(sid)

         if not user_id:
@@ -89,7 +89,7 @@ def create_app():

     @socketio.on("disconnect")
     def handle_disconnect():
-        sid = socketio.server.eio.sid
+        sid = request.sid
         if sid in connected_users:
             user_id = connected_users.pop(sid)
             print(f"Client disconnected with ID: {sid}, User ID: {user_id}")
@@ -102,7 +102,7 @@ def create_app():
     @socketio.on("message")
     def handle_message(data):
         try:
-            sid = socketio.server.eio.sid
+            sid = request.sid
             user_id = connected_users.get(sid)
             if not user_id:
                 print(f"No authenticated user found for message from {sid}")
```

### NOT work

The client is rejected by the server :(

```json
{"message":"Connection rejected by server"}
```

## Where in documentation

I searched in the documentation about `request.sid` but I didn't find anything.

[flask-socketio: search "request.sid"](https://flask-socketio.readthedocs.io/en/latest/search.html?q=request.sid)

## Werkzeug: `assert status_set`

It seems normal in development mode from this comment https://github.com/miguelgrinberg/flask-sock/issues/27#issuecomment-1121972991

```
Error on request:
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/werkzeug/serving.py", line 370, in run_wsgi
    execute(self.server.app)
  File "/usr/local/lib/python3.9/site-packages/werkzeug/serving.py", line 336, in execute
    write(b"")
  File "/usr/local/lib/python3.9/site-packages/werkzeug/serving.py", line 261, in write
    assert status_set is not None, "write() before start_response"
AssertionError: write() before start_response
```

**Comment by miguelgrinberg** [comment]
You have to explain better what you are doing. Flask-SocketIO is a server, but your non-working error appears to indicate you are doing things in the client that fail.

As far as documentation, searching for `sid` should lead you to https://flask-socketio.readthedocs.io/en/latest/implementation_notes.html.

The correct way to get the `sid` is with `request.sid`. The other expression that you have is not what you want as it uses private parts of this package that are not exposed to applications.

**Comment by t-h2o** [comment]
Thanks.

> You have to explain better what you are doing.

My goal is to implement socketIO in my Flask project.
I opened this discussion because after applying, the server behavior has changed.

> your non-working error appears to indicate you are doing things in the client that fail.

Yes, non-working error are in the client, I use it to test the server.

- working: after the login I receive a SID.
- non-working: after the login, I receive this error "Connection rejected by server".

> The other expression that you have is not what you want as it uses private parts of this package that are not exposed to applications.

Now I understand why I couldn't find anything about `SocketIO.server` in [API Reference](https://flask-socketio.readthedocs.io/en/latest/api.html).

I call in my program `SocketIO.server`. How can I get the environ without using `SocketIO.server` it it is private?

```python
# Get environment for this connection
environ = socketio.server.environ.get(sid, {})

# Check auth data
auth = environ.get("socketio", {}).get("auth", {})
```

**Reply by miguelgrinberg** [reply]
Anything that is not in the documentation is private and should not be used. If you use these things then your application may break when you upgrade to a newer version, because I cannot guarantee that private elements will never change.

I still don't understand what you are trying to do, but I think you are using the `sid` for things that you shouldn't. The `sid` is just a unique code that identifies a session. For a given client, there is a `sid` for the low-level transport, and a `sid` for each connected namespace. These are all different, and it is likely that you are mixing them up.

Also note that flask provides access to the `environ` for your client with `request.environ`.
